src/scales/scale.time.js

```diff
@@ -61,253 +69,288 @@ module.exports = function(Chart) {
+		return moment(label, timeOpts.format);
+	}
+
+	/**
+	 * Figure out which is the best unit for the scale
+	 * @param minUnit {String} minimum unit to use
+	 * @param min {Number} scale minimum
+	 * @param max {Number} scale maximum
+	 * @return {String} the unit to use
+	 */
+	function determineUnit(minUnit, min, max, maxTicks) {
+		var units = Object.keys(time);
+		var unit;
+
+		for (var i = units.indexOf(minUnit); i < units.length; i++) {
+			unit = units[i];
+			var unitDetails = time[unit];
+			var fits = false;
+
+			if (unitDetails.steps) {
+				fits = Math.ceil((max - min) / (unitDetails.steps[unitDetails.steps.length - 1] * unitDetails.size)) <= maxTicks;
+			} else if (unitDetails.maxStep) {
+				fits = Math.ceil((max - min) / (unitDetails.maxStep * unitDetails.size)) <= maxTicks;
+			} else {
+				// No limit on the multiplier so it always fits
+				fits = true;
 			}
 
-			Chart.Scale.prototype.initialize.call(this);
-		},
-		getLabelDiff: function(datasetIndex, index) {
-			var me = this;
-			if (datasetIndex === null || index === null) {
-				return null;
+			if (fits) {
+				break;
 			}
+		}
 
-			if (me.labelDiffs === undefined) {
-				me.buildLabelDiffs();
+		return unit;
+	}
+
+	/**
+	 * Determines how we scale the unit
+	 * @param min {Number} the scale minimum
+	 * @param max {Number} the scale maximum
+	 * @param unit {String} the unit determined by the {@see determineUnit} method
+	 * @return {Number} the axis step size in milliseconds
+	 */
+	function determineStepSize(min, max, unit, maxTicks) {
+		// Using our unit, figoure out what we need to scale as
+		var unitDefinition = time[unit];
+		var unitSizeInMilliSeconds = unitDefinition.size;
+		var sizeInUnits = Math.ceil((max - min) / unitSizeInMilliSeconds);
+		var multiplier = 1;
+
+		if (unitDefinition.steps) {
+			// Have an array of steps
+			for (var i = 0; i < unitDefinition.steps.length && sizeInUnits > maxTicks; i++) {
+				multiplier = unitDefinition.steps[i];
+				sizeInUnits = Math.ceil((max - min) / (unitSizeInMilliSeconds * multiplier));
 			}
+		} else {
+			while (sizeInUnits > maxTicks) {
+				++multiplier;
+				sizeInUnits = Math.ceil((max - min) / (unitSizeInMilliSeconds * multiplier));
+			}
+		}
 
-			if (typeof me.labelDiffs[datasetIndex] !== 'undefined') {
-				return me.labelDiffs[datasetIndex][index];
+		return multiplier;
+	}
+
+	/**
+	 * Helper for generating axis labels.
+	 * @param generationOptions {ITimeGeneratorOptions} the options for generation
+	 * @param dataRange {IRange} the data range
+	 * @param niceRange {IRange} the pretty range to display
+	 * @return {Number[]} ticks
+	 */
+	function generateTicks(generationOptions, dataRange, niceRange) {
+		var ticks = [];
+		var stepSize = generationOptions.stepSize;
+		ticks.push(generationOptions.min !== undefined ? generationOptions.min : niceRange.min);
+		var cur = moment(niceRange.min);
+		while (cur.add(stepSize, generationOptions.unit).valueOf() < niceRange.max) {
```

Comment:
This simplifies units which don't have a constant number of milliseconds in them eg months, years.